dataverse-webapp/src/test/java/edu/harvard/iq/dataverse/workflow/artifacts/WorkflowArtifactServiceBeanTest.java

```diff
@@ -0,0 +1,185 @@
+        assertThat(artifact.getName()).isEqualTo(NAME);
+        assertThat(artifact.getEncoding()).isEqualTo(ENCODING);
+
+        assertThat(storageService.getNumberOfStoredObjects()).isEqualTo(1);
+    }
+
+    @Test
+    @DisplayName("Should be able to retrieve stored artifact data")
+    public void shouldRetrieveStoredData() throws IOException {
+        // given
+        ArtifactData data = new ArtifactData(NAME, ENCODING, TEST_DATA_SUPPLIER);
+
+        // when
+        WorkflowArtifact artifact = serviceBean.saveArtifact(1L, data);
+        Optional<Supplier<InputStream>> streamSupplier = serviceBean.readAsStream(artifact);
+
+        // then
+        assertThat(streamSupplier.isPresent()).isTrue();
+
+        assertThat(storageService.getReadCounter()).isEqualTo(1);
+    }
+
+    @Test
+    @DisplayName("Should delete stored artifacts for the given dataset version")
+    public void shouldDeleteArtifactsForGivenDatasetVersion() {
+        // given
+        final int size = 10;
+        ArtifactData data = new ArtifactData(NAME, ENCODING, TEST_DATA_SUPPLIER);
+
+        // when
+        IntStream.rangeClosed(1, size).forEach(i -> serviceBean.saveArtifact(1L, data));
+        List<WorkflowArtifact> before = repository.findAllByWorkflowExecutionId(1L);
+        serviceBean.deleteArtifacts(1L);
+        List<WorkflowArtifact> after = repository.findAllByWorkflowExecutionId(1L);
+
+        // then
+        assertThat(before.size()).isEqualTo(size);
+        assertThat(after).isEmpty();
+
+        assertThat(storageService.getNumberOfStoredObjects()).isEqualTo(0);
+    }
+
+    // -------------------- INNER CLASSES --------------------
+
+    private static class TestStorageService implements StorageService {
+        private Map<String, Supplier<InputStream>> storage = new HashMap<>();
+
+        private int readCounter = 0;
+
+        @Override
+        public StorageType getStorageType() {
+            return StorageType.DATABASE;
+        }
+
+        @Override
+        public String save(Supplier<InputStream> inputStreamSupplier) {
+            String location = UUID.randomUUID().toString();
+            storage.put(location, inputStreamSupplier);
+            return location;
+        }
+
+        @Override
+        public Optional<Supplier<InputStream>> readAsStream(String location) {
+            readCounter++;
+            Supplier<InputStream> inputStreamSupplier = storage.get(location);
+            return inputStreamSupplier != null
+                    ? Optional.of(inputStreamSupplier)
+                    : Optional.empty();
+        }
+
+        @Override
+        public void delete(String location) {
+            storage.remove(location);
+        }
+
+        public int getNumberOfStoredObjects() {
+            return storage.size();
+        }
+
+        public int getReadCounter() {
+            return readCounter;
+        }
+    }
+
+    @Vetoed
+    private static class Repository extends WorkflowArtifactRepository {
```

Comment:
Possibly, but it would require creating additional interface for this repo and repeating at least a half of the code below, not counting mock configuration etc., so I think that it would be better to leave this as it is now.